src/bootstrap/doc.rs

```diff
@@ -563,6 +563,80 @@ impl Step for Test {
     }
 }
 
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct ProcMacro {
+    stage: u32,
+    target: Interned<String>,
+}
+
+impl Step for ProcMacro {
+    type Output = ();
+    const DEFAULT: bool = true;
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun) -> ShouldRun {
+        let builder = run.builder;
+        run.krate("rustc-main").default_condition(builder.build.config.docs)
+    }
+
+    fn make_run(run: RunConfig) {
+        run.builder.ensure(ProcMacro {
+            stage: run.builder.top_stage,
+            target: run.target,
+        });
+    }
+
+    /// Generate proc_macro documentation.
+    ///
+    /// This will generate all documentation for procedural macros and
+    /// dependencies. Procedural macro documentation is included in the
+    /// standard Rust documentation, so we should always document it and
+    /// symlink to merge with the rest of the std and test documentation.
+    /// We don't build other compiler documentation here as we want to be
+    /// able to keep it separate from the standard documentation.
+    /// This is largely just a wrapper around `cargo doc`.
+    fn run(self, builder: &Builder) {
+        let build = builder.build;
+        let stage = self.stage;
+        let target = self.target;
+        println!("Documenting stage{} procedural macro ({})", stage, target);
+        let out = build.doc_out(target);
+        t!(fs::create_dir_all(&out));
+        let compiler = builder.compiler(stage, build.build);
+        let rustdoc = builder.rustdoc(compiler.host);
+        let compiler = if build.force_use_stage1(compiler, target) {
+            builder.compiler(1, compiler.host)
+        } else {
+            compiler
+        };
+
+        // Build libstd docs so that we generate relative links
+        builder.ensure(Std { stage, target });
+
+        builder.ensure(compile::Rustc { compiler, target });
+        let out_dir = build.stage_out(compiler, Mode::Librustc)
+                           .join(target).join("doc");
+
+        // See docs in std above for why we symlink
+        let my_out = build.crate_doc_out(target);
+        build.clear_if_dirty(&my_out, &rustdoc);
+        t!(symlink_dir_force(&my_out, &out_dir));
+
+        let mut cargo = builder.cargo(compiler, Mode::Librustc, target, "doc");
+        compile::rustc_cargo(build, &mut cargo);
+
+        // Like with libstd above if compiler docs aren't enabled then we're not
+        // documenting internal dependencies, so we have a whitelist.
+        cargo.arg("--no-deps");
+        for krate in &["proc_macro"] {
```

Comment:
... why? 😕

Reply:
@davidtwco  Sorry, let me clarify.

Why iterate on an array with a single element when you could just write `cargo.arg("-p").arg("proc_macro")` directly? The step name doesn't suggest you'll need to extend this array.

Reply:
Before this PR, when compiler docs were disabled, a whitelist of crates from the compiler were documented for the standard documentation. This was clear from the context and that it was an array. However, as you point out, that context has been lost somewhat.

I had mentioned in my initial comment that I wasn't sure about the name of this step. I named it `ProcMacro` just so that it had a name for me to work with, in practice it should be `PublicRustc` or `WhitelistRustc` or something like that since it could contain more crates that we choose to include in the standard available documentation.

I'm not sure what name is best for this step but I'll change it and clarify the point of it in a comment.

Reply:
Modified that commit to clarify this.